File: shiw/shiw/report/daily_heat_loss_report/daily_heat_loss_report.js
// Copyright (c) 2025, beetashoke chakraborty and contributors
// For license information, please see license.txt

// frappe.query_reports["Daily Heat Loss report"] = {
// 	"filters": [

// 	]
// };



frappe.query_reports["Daily Heat Loss report"] = {
	filters: [
		{
			fieldname: "from_date",
			label: "From Date",
			fieldtype: "Date",

		},
		{
			fieldname: "to_date",
			label: "To Date",
			fieldtype: "Date",

		},
		{
			fieldname: "shift_type",
			label: "Shift Type",
			fieldtype: "Select",
			options: "\nDay shift\nNight shift"
		},
		{
			fieldname: "reason_for_heat_loss",
			label: "Reason for Heat Loss",
			fieldtype: "Select",
			options: [
				"",
				"Over Achieved",
				"Manpower Issue",
				"Power Problem",
				"Ladle Not Available",
				"No Mould in HPML",
				"Stopper Jam",
				"Chemistry Out",
				"Grade Change",
				"Rusted Scrap Boiling Heat Slow Running",
				"Sintering Heat",
				"Ladle Breakdown",
				"Cold Furnace",
				"Sampling",
				"Patching Heat",
				"Lack Of Raw Material",
				"High Water Temp",
				"Furnace Repairing",
				"Bariman issue",
				"Power cut",
				"No mould in JS",
				"Crane man unavailable",
				"Scrap problem",
				"Pattern change",
				"Heat change",
				"Nozzle Leakage",
				"Due to both 1 ton furnaces running at the same time",
				"Furnace starts late",
				"Furnace hold",
				"Nozzle Break",
				"SS Heat take time",
				"Furnace Breakdown",
				"DI Heat Change",
				"Chemistry High",
				"Mixed scrap",
				"Ladle heating issue",
				"Furnace stop",
				"Grade change"
			].join("\n")
		}
	]
};

File: shiw/shiw/report/daily_heat_loss_report/daily_heat_loss_report.py
# Copyright (c) 2025, beetashoke chakraborty and contributors
# For license information, please see license.txt

# import frappe


# def execute(filters=None):
# 	columns, data = [], []
# 	return columns, data




import frappe
from frappe.utils import nowdate

def execute(filters=None):
    if not filters:
        filters = {}

    from_date = filters.get("from_date")
    to_date = filters.get("to_date")
    shift_type = filters.get("shift_type")
    reason = filters.get("reason_for_heat_loss")

    conditions = ["parent.docstatus = 1"]
    values = {}

    # If dates provided, use them; else default to all till today
    if from_date and to_date:
        conditions.append("parent.date BETWEEN %(from_date)s AND %(to_date)s")
        values["from_date"] = from_date
        values["to_date"] = to_date
    else:
        # Show all till today
        conditions.append("parent.date <= %(today)s")
        values["today"] = nowdate()

    if shift_type:
        conditions.append("parent.shift_type = %(shift_type)s")
        values["shift_type"] = shift_type

    if reason:
        conditions.append("child.reason_for_heat_loss = %(reason)s")
        values["reason"] = reason

    condition_str = " AND ".join(conditions)

    query = f"""
        SELECT
            child.reason_for_heat_loss,
            SUM(child.weight_in_kg) AS total_weight
        FROM
            `tabReason For Heat Loss Table` AS child
        INNER JOIN
            `tabDaily Heat Loss` AS parent ON child.parent = parent.name
        WHERE {condition_str}
        GROUP BY
            child.reason_for_heat_loss
        ORDER BY
            total_weight DESC
    """

    data = frappe.db.sql(query, values, as_dict=True)

    columns = [
        {"label": "Reason for Heat Loss", "fieldname": "reason_for_heat_loss", "fieldtype": "Data", "width": 400},
        {"label": "Total Weight (kg)", "fieldname": "total_weight", "fieldtype": "Float", "width": 150},
    ]

    return columns, data
